src/checkpoint.hh

```diff
@@ -0,0 +1,1246 @@
+
+namespace Printer {
+class Checkpoint : public Base {
+ private:
+  // true if table types are wrapped in ListRef
+  bool list_ref;
+
+  std::vector<std::string> string_type_accessors, block_type_accessors,
+                           subseq_type_accessors;
+
+  // currently supported/serializable datatypes
+  const std::array<Type::Type, 11>
+  SUPPORTED_TYPES = {Type::Type::VOID, Type::Type::INTEGER,
+                     Type::Type::INT, Type::Type::FLOAT,
+                     Type::Type::SIZE, Type::Type::SINGLE,
+                     Type::Type::BIGINT, Type::Type::STRING,
+                     Type::Type::SHAPE, Type::Type::SUBSEQ,
+                     Type::Type::EXTERNAL};
+
+const std::array<std::string, 12>
+SUPPORTED_EXTERNAL_TYPES = {"Rope", "answer_pknot_mfe", "pktype",
+                            "answer_pknot_mfecovar", "mfecovar",
+                            "mfecovar_macrostate", "pftuple",
+                            "answer_pknot_pfunc",
+                            "answer_ali_pfunc_macrostate",
+                            "answer_macrostate_mfe",
+                            "answer_macrostate_pfunc", "shape_t"};
+
+  template<size_t n>
+  bool has_type(const std::array<Type::Type, n> &arr, Type::Type type,
+                Type::Base *t) {
+    for (Type::Type supported_type : arr) {
+      if (type == supported_type) {
+        if (type == Type::Type::STRING) {
+          strings = true;
+        } else if (type == Type::Type::SUBSEQ) {
+          subseq = true;
+        } else if (type == Type::Type::EXTERNAL) {
+          // only allow external type "Rope" and some fold-grammars types
+          Type::External *e = dynamic_cast<Type::External*>(t);
+          bool supported_external_type = false;
+          for (const std::string &supported_ext : SUPPORTED_EXTERNAL_TYPES) {
+            if (*e->name == supported_ext) {
+              supported_external_type = true;
+              break;
+            }
+          }
+          if (!supported_external_type) return false;
+        }
+          return true;
+      }
+    }
+    return false;
+  }
+
+  void add_type_accessor(Type::Type type,
+                         std::vector<std::string> &type_accessor) {
+    if (type == Type::Type::STRING) {
+      std::stringstream final_type_accessor;
+      for (auto el : type_accessor) final_type_accessor << el;
+      string_type_accessors.push_back(final_type_accessor.str() +
+                                      ".block_as_int");
+      block_type_accessors.push_back(final_type_accessor.str() +
+                                     ".get_block()");
+    } else if (type == Type::Type::SUBSEQ) {
+      std::stringstream final_type_accessor;
+      for (auto el : type_accessor) final_type_accessor << el;
+      subseq_type_accessors.push_back(final_type_accessor.str() +
+                                      ".seq");
+    }
+  }
+
+  bool __is_supported(Type::Base *type) {
```

Comment:
what is the meaning of the leading `__` ?

Reply:
No special meaning, I just use this as an indicator that this is a private/internal function not meant to be called by the user (which isn't permitted anyways since it is private).